### ISSUE TITLE
DesktopHeapType.FixGenerics - does not work when subtype arity is missing or invalid

### ISSUE THREAD

**Issue by holc-tkomp** [microsoft/clrmd#394]
I've noticed that [`FixGenericsWorker`](https://github.com/microsoft/clrmd/blob/66687332950808cf49334a14814a2158852aa43d/src/Microsoft.Diagnostics.Runtime/src/Desktop/DesktopHeapType.cs#L840) heavily relies on the subtype arity (number of type parameters) on deciding how to process the name.

ECMA-335 (I.10.7.2 Type names and arity encoding) states that `the grave accent character and arity together are optional`, but that is followed by the `Rule 43` which states slightly different version 😕.

**Using arity alone to decide whether the name is a generic type and getting count of type parameters doesn't seem right. It should be optional and so work properly even if it's not present.**

Omitting the subtype arity, does not cause issues in runtime, but memory dumps will not contain the arity, therefore, given the types:
```
1. MyAssembly.Test`1[[System.String, mscorlib]]
2. MyAssembly.Test[[System.String, mscorlib]]
3. MyAssembly.Test`3[[System.String, mscorlib]]
```
`ClrType`s will have the `Name`s set to:
```
1. MyAssembly.Test<System.String>
2. MyAssembly.Test[[System.String, mscorlib]]
3. MyAssembly.Test<System.String,
```

**Comment by leculver** [comment]
Does the third variety ever happen in practice?  If the arity is present it should match the number of parameters right?

**Comment by holc-tkomp** [comment]
You're right. I unnecessarily pasted this example, it probably never happens and is invalid according to that ECMA.

The 2nd example is what bothers me and my company's software.

**Comment by leculver** [comment]
I've fixed the issue in the ClrMD 2.0 branch.  I'm wary of shipping it as an update to clrmd 1.1 without a lot of additional testing.  That style of manual text parsing is very error prone and every time I make changes it always seems to break in some rare corner case I didn't consider.  I'll likely make a very risk mitigating change when I port it over.

I'm keeping this issue open until I fix it in 1.1.
